[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66295
* #65344
* #65178
* #65177
* __->__ #65122
* #65343

Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

swolchok committed:
[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations

Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

[ghstack-poisoned]

swolchok committed:
Update on "[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations"


Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

[ghstack-poisoned]

eellison commented:
will get to this tomorrow..

eellison commented on torch/csrc/jit/passes/utils/memory_dag.cpp:
```diff
@@ -189,6 +203,7 @@ void MemoryDAG::setWildcards(
   // For every element, if the cache contains `MemoryLocationFoo`, then we must
   // add `WildcardBar` to it.
   for (const std::unique_ptr<Element>& e : this->indexToElementMap_) {
+    e->cachedAllContainedMemoryLocations_.reset();
```
Why do we need to do this ?

swolchok replied:
we're mutating the graph, so we have to invalidate the caches (or update them, but that seems difficult to do efficiently)

swolchok committed:
Update on "[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations"


Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

[ghstack-poisoned]

swolchok commented on c10/util/sparse_bitset.h:
```diff
@@ -548,6 +548,11 @@ class SparseBitVector {
       return false;
 
+    if (empty()) {
+      *this = RHS;
```
it's an assignment operator call -- delegating to the assignment operator is a bit more efficient

swolchok committed:
Update on "[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations"


Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

[ghstack-poisoned]

swolchok commented:
ping

swolchok commented:
ping @eellison

swolchok committed:
Update on "[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations"


Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

[ghstack-poisoned]

eellison reviewed (commented):
LGTM. The one sort of worry I had is it's interaction with our public mutating apis `replaceWithNewValue`, `copyValue`, `createValue`, but all of those apis retain the validity of the memory dag and strictly change the mapping from Value * -> Element *. I would still like to wait on @suo's review and it would also be helpful to add a test or two for the invalidation.

suo reviewed (approved):
This looks fine to me. lgtm as well. The cache invalidation is a touch conservative even—but probably not worth making more accurate.

swolchok committed:
Update on "[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations"


Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

[ghstack-poisoned]

swolchok committed:
Update on "[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations"


Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

[ghstack-poisoned]
